crash in session_test.go if datacenter is not expected name

Related to #1365 

=== RUN   TestSessionDefaults
--- PASS: TestSessionDefaults (0.42s)
	session_test.go:46: &{Client:0xc8201d1590 Config:0xc8200814a0 Cluster:ClusterComputeResource:domain-c7 @ /Datacenter/host/TestCluster Datacenter:Datacenter:datacenter-2 @ /Datacenter Datastore:Datastore:datastore-27 @ /Datacenter/datastore/datastore1 Host:HostSystem:host-26 @ /Datacenter/host/TestCluster/10.118.67.103 Pool:ResourcePool:resgroup-8 @ /Datacenter/host/TestCluster/Resources Finder:0xc8201de420 folders:<nil>}
=== RUN   TestSession
--- FAIL: TestSession (0.11s)
	session_test.go:64: Failure finding ds (/ha-datacenter/datastore/*): datastore '/ha-datacenter/datastore/*' not found
		Failure finding host (/ha-datacenter/host/*/*): host '/ha-datacenter/host/*/*' not found
		Failure finding pool (/ha-datacenter/host/*/Resources): resource pool '/ha-datacenter/host/*/Resources' not found
	session_test.go:66: Failure finding ds (/ha-datacenter/datastore/*): datastore '/ha-datacenter/datastore/*' not found
		Failure finding host (/ha-datacenter/host/*/*): host '/ha-datacenter/host/*/*' not found
		Failure finding pool (/ha-datacenter/host/*/Resources): resource pool '/ha-datacenter/host/*/Resources' not found
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x48c23e]

goroutine 10 [running]:
panic(0xf29e00, 0xc8200121a0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
testing.tRunner.func1(0xc82042e090)
	/usr/local/go/src/testing/testing.go:467 +0x192
panic(0xf29e00, 0xc8200121a0)
	/usr/local/go/src/runtime/panic.go:443 +0x4e9
github.com/vmware/vic/pkg/vsphere/session.TestSession(0xc82042e090)
	/drone/src/github.com/vmware/vic/pkg/vsphere/session/session_test.go:71 +0x3be
testing.tRunner(0xc82042e090, 0x1bdc1f8)
	/usr/local/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0x892
exit status 2
FAIL	github.com/vmware/vic/pkg/vsphere/session	0.546s
make: *** [test] Error 1
Makefile:205: recipe for target 'test' failed
[info] build failed (exit code 2)

--- a/pkg/vsphere/session/session_test.go
+++ b/pkg/vsphere/session/session_test.go
@@ -15,6 +15,7 @@
 package session
 
 import (
+	"strings"
 	"testing"
 	"time"
 
@@ -41,7 +42,9 @@ func TestSessionDefaults(t *testing.T) {
 			t.SkipNow()
 		}
 	}
-	defer session.Logout(ctx)
+	if session != nil {
+		defer session.Logout(ctx)
+	}
 
 	t.Logf("%+v", session)
 }
@@ -61,19 +64,30 @@ func TestSession(t *testing.T) {
 
 	session, err := NewSession(config).Create(ctx)
 	if err != nil {
-		t.Logf("%+v", err.Error())
+		eStr := err.Error()
+		t.Logf("%+v", eStr)
+		// FIXME: session.Create incorporates Populate which loses the type of any original error from vmomi
+		// In the case where the test is run on a cluster with multiple hosts, find.MultipleFoundError
+		// gets rolled up into a generic error in Populate. As such, the best we can do is just grep for the string, which is lame
+		// The test shouldn't fail if it's run on a cluster with multiple hosts. However, it won't test for anything either.
+		if strings.Contains(eStr, "resolves to multiple hosts") {
+			t.SkipNow()
+		}
 		if _, ok := err.(*find.MultipleFoundError); !ok {
-			t.Errorf(err.Error())
+			t.Errorf(eStr)
 		} else {
 			t.SkipNow()
 		}
 	}
-	defer session.Logout(ctx)
 
-	t.Logf("Session: %+v", session)
+	if session != nil {
+		defer session.Logout(ctx)
 
-	t.Logf("IsVC: %t", session.IsVC())
-	t.Logf("IsVSAN: %t", session.IsVSAN(ctx))
+		t.Logf("Session: %+v", session)
+
+		t.Logf("IsVC: %t", session.IsVC())
+		t.Logf("IsVSAN: %t", session.IsVSAN(ctx))
+	}
 }
 
 func TestFolder(t *testing.T) {
@@ -91,17 +105,25 @@ func TestFolder(t *testing.T) {
 
 	session, err := NewSession(config).Create(ctx)
 	if err != nil {
-		t.Logf("%+v", err.Error())
+		eStr := err.Error()
+		t.Logf("%+v", eStr)
+		// FIXME: See comments above
+		if strings.Contains(eStr, "resolves to multiple hosts") {
+			t.SkipNow()
+		}
 		if _, ok := err.(*find.MultipleFoundError); !ok {
-			t.Errorf(err.Error())
+			t.Errorf(eStr)
 		} else {
 			t.SkipNow()
 		}
 	}
-	defer session.Logout(ctx)
 
-	folders := session.Folders(ctx)
-	if folders == nil || folders.VmFolder == nil {
-		t.Errorf("Get empty folder")
+	if session != nil {
+		defer session.Logout(ctx)
+
+		folders := session.Folders(ctx)
+		if folders == nil || folders.VmFolder == nil {
+			t.Errorf("Get empty folder")
+		}
 	}
 }

